### PR TITLE
[DED-230] - Add Bookmarks component, react-query, logic around that

### PR DESCRIPTION
**Pull request by mpsujek**
# Deploy preview

Scroll down to the checks section for the link to the deploy preview of your branch.

# Links

Link your Jira ticket.

- [Jira](https://livechatinc.atlassian.net/browse/DED-230)

# Description

# Review and release

Apply changes and resolve conflicts. Once the described functionality lands on prod, let us know on #platform-docs-releases that your PR is ready for release. We will **merge** and **publish** the changes.

### CODE REVIEW

File: src/components/core/Bookmark/loginmodal.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/** @jsx jsx */ import { jsx, css } from "@emotion/core";
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/loginmodal.js:1]: Please rename the filename with React files naming convention - should be `LoginModal.js`

File: src/components/core/Bookmark/BookmarkIcon.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  align-items: center;
+  border-radius: 50%;
+  :hover {
+    background-color: #f6f6f7;
+  }
+`;
+
+const BookmarkIcon = ({ isBookmarked, handleClick }) => {
+  return (
+    <Container onClick={handleClick}>
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/BookmarkIcon.js:25]: I'm not sure about this component from the semantic side, as div shouldn't be used as a button

File: src/components/core/Bookmark/index.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+import React, { useState, useEffect } from "react";
+import { useQuery } from "react-query";
+import { useAuth } from "../../../contexts/auth";
+import { useRouter } from "next/router";
+import { useUpdateBookmarks } from "hooks";
+import BookmarkIcon from "./BookmarkIcon";
+import LoginModal from "./loginmodal";
+import api from "api";
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/index.js:8]: You should sort imports

File: src/components/core/Bookmark/index.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        return false;
+      });
+    }
+  });
+
+  return (
+    <>
+      <BookmarkIcon
+        isBookmarked={isBookmarked}
+        handleClick={() => handleBookmarkClick()}
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/index.js:68]: Simpler form - `handleClick={handleBookmarkClick}`

File: src/components/core/Bookmark/index.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  const { isLoading, data: allBookmarks } = useQuery(
+    ["allBookmarks"],
+    fetchData
+  );
+
+  const newBookmark = [{ type: "bookmark", url: pagePath }];
+
+  const handleBookmarkClick = () => {
+    if (isAuthorized) {
+      fetchData();
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/index.js:33]: Do we need this call?

File: src/components/core/Bookmark/index.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        const filteredBookmarks = Object.values(allBookmarks).filter(
+          (singleBookmark) => singleBookmark.url !== pagePath
+        );
+
+        updateBookmarks(filteredBookmarks);
+      } else {
+        setBookmark(true);
+        const updatedBookmarks = [...allBookmarks, ...newBookmark];
+        updateBookmarks(updatedBookmarks);
+      }
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/index.js:45]: How would you know that something is already bookmarked from the backend data perspective? 

It seems that you mark that something is bookmarked only after user click it before in this session

Comment (SamborGornicz) [reply src/components/core/Bookmark/index.js:45]: nevermind I've checked the rest of the code 😄

File: src/components/core/Bookmark/index.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import { useQuery } from "react-query";
+import { useAuth } from "../../../contexts/auth";
+import { useRouter } from "next/router";
+import { useUpdateBookmarks } from "hooks";
+import BookmarkIcon from "./BookmarkIcon";
+import LoginModal from "./loginmodal";
+import api from "api";
+
+const Bookmark = () => {
+  const [isBookmarked, setBookmark] = useState("");
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/index.js:11]: Should be `setIsBookmarked`

Comment (SamborGornicz) [reply src/components/core/Bookmark/index.js:11]: and default value should be `false`

File: src/components/core/Bookmark/index.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    const bookmarks = await api.getDPS().getAllBookmarks();
+    return bookmarks;
+  };
+
+  const { isLoading, data: allBookmarks } = useQuery(
+    ["allBookmarks"],
+    fetchData
+  );
+
+  const newBookmark = [{ type: "bookmark", url: pagePath }];
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/index.js:29]: We could put it inside of `handleBookmarkClick` function, as we don't need it elsewhere

File: src/components/core/Bookmark/index.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      setModalOpen(true);
+    }
+  };
+
+  useEffect(() => {
+    if (!isLoading && isAuthorized) {
+      Object.values(allBookmarks).some((element) => {
+        if (element.url === pagePath) {
+          setBookmark(true);
+          return true;
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/index.js:56]: You don't need to return here

File: src/components/core/Bookmark/index.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        setBookmark(false);
+        return false;
+      });
+    }
+  });
```
Comment (SamborGornicz) [inline src/components/core/Bookmark/index.js:62]: Lack of dependencies array, without this, component will run this function after each rerender

File: src/components/core/Bookmark/BookmarkIcon.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      {isBookmarked ? <BookmarkFilledIcon /> : <BookmarkHollowIcon />}
+    </Container>
+  );
+};
+
+BookmarkIcon.propTypes = {
+  handleClick: func.isRequired,
```
Comment (arturfracala) [inline src/components/core/Bookmark/BookmarkIcon.js:32]: You're not checking `isBookmarked ` props.

File: src/components/core/Bookmark/BookmarkIcon.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  border-radius: 50%;
+  :hover {
+    background-color: #f6f6f7;
+  }
+`;
+
+const BookmarkIcon = ({ isBookmarked, handleClick }) => {
```
Comment (arturfracala) [inline src/components/core/Bookmark/BookmarkIcon.js:23]: I don't like this approach. You have a component called `BookmarkIcon`, but in fact, it's not an icon. Icons should be placed in the `assets` directory. I wouldn't put this logic in a separate directory imho.

File: src/components/core/Bookmark/index.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+import { useRouter } from "next/router";
+import { useUpdateBookmarks } from "hooks";
+import BookmarkIcon from "./BookmarkIcon";
+import LoginModal from "./loginmodal";
+import api from "api";
+
+const Bookmark = () => {
+  const [isBookmarked, setBookmark] = useState("");
+  const { authorize, isAuthorized, user } = useAuth();
+  const [isModalOpen, setModalOpen] = useState(false);
```
Comment (arturfracala) [inline src/components/core/Bookmark/index.js:13]:
```suggestion
  const [isModalOpen, setIsModalOpen] = useState(false);
```
Keep the naming convention of `useState` hook.

File: src/components/core/Bookmark/index.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+  const router = useRouter();
+  const pagePath = router.pathname;
```
Comment (arturfracala) [inline src/components/core/Bookmark/index.js:15]:
```suggestion
  const { pathname } = useRouter();
```
Maybe this way would be easier.

File: src/components/core/Bookmark/index.js [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+  );
+
+  const newBookmark = [{ type: "bookmark", url: pagePath }];
+
+  const handleBookmarkClick = () => {
+    if (isAuthorized) {
+      fetchData();
+      if (isBookmarked) {
+        setBookmark(false);
+        const filteredBookmarks = Object.values(allBookmarks).filter(
```
Comment (arturfracala) [inline src/components/core/Bookmark/index.js:36]: This line gonna be a problem if someone doesn't have any shortcuts. There will be `TypeError: Cannot convert undefined or null to object` error.

File: src/components/core/Bookmark/index.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        updateBookmarks(updatedBookmarks);
+      }
+    } else {
+      setModalOpen(true);
+    }
+  };
+
+  useEffect(() => {
+    if (!isLoading && isAuthorized) {
+      Object.values(allBookmarks).some((element) => {
```
[the remark this answers is not in the document]
Comment (arturfracala) [reply src/components/core/Bookmark/index.js:53]: This line gonna be a problem if someone doesn't have any shortcuts. There will be TypeError: Cannot convert undefined or null to object error.

File: src/components/core/Bookmark/loginmodal.js [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+  }
+`;
+
+const ButtonsWrapper = styled.div`
+  width: 35%;
+  display: flex;
+  justify-content: space-between;
+`;
+
+const LoginModal = ({ authorize, isOpen, handleModalClose, setBookmark }) => {
```
Comment (arturfracala) [inline src/components/core/Bookmark/loginmodal.js:34]: Check props types

File: src/components/core/Bookmark/loginmodal.js [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+  return (
+    <ModalBase onClose={onModalClose} css={modalBaseCss}>
+      <LoginIcon />
+      <h2>Log in to add bookmarks</h2>
+      <p>You need to be logged in to bookmark documents.</p>
+      <ButtonsWrapper>
+        <Button
+          onClick={() => {
+            handleModalClose();
+          }}
```
Comment (arturfracala) [inline src/components/core/Bookmark/loginmodal.js:56]:
```suggestion
          onClick={handleModalClose}
```

File: src/components/core/Bookmark/loginmodal.js [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+          kind="secondary"
+        >
+          Cancel
+        </Button>
+        <Button
+          onClick={() => {
+            onLoginProceed();
+          }}
```
Comment (arturfracala) [inline src/components/core/Bookmark/loginmodal.js:64]:
```suggestion
          onClick={onLoginProceed}
```

File: src/components/core/Bookmark/loginmodal.js [hunk 20]
```diff
@@ -0,0 +1,74 @@
+/** @jsx jsx */ import { jsx, css } from "@emotion/core";
+import styled from "@emotion/styled";
+import { Button, ModalBase } from "@livechat/design-system";
+import { LoginIcon } from "assets/icons/LoginIcon";
+import { logAmplitudeEvent } from "../../../utils";
```
Comment (arturfracala) [inline src/components/core/Bookmark/loginmodal.js:5]: Put a blank line above of the nested import

File: package.json [hunk 21]
```diff
@@ -71,11 +72,11 @@
   },
   "devDependencies": {
     "@netlify/plugin-nextjs": "^3.9.2",
+    "husky": "^8.0.0",
```
Comment (arturfracala) [inline package.json:75]: Why did you change the dependencies order?

**Review by arturfracala — CHANGES_REQUESTED** [review]
Are you sure you installed node modules with the correct node version?

File: package.json (hunk 21 above)
Comment (mpsujek) [reply package.json:75]: To be honest I didn't see that was changed. I'll revert that.

File: src/api/dps.js [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    });
+  }
+
+  async getAllBookmarks() {
+    try {
+      const res = await this.instance.get(`/v2/links`);
+      return res.data;
+    } catch (error) {
+      console.log(error);
+      throw new Error(`Can't get bookmarks [${error.message}]`);
```
Comment (arturfracala) [inline src/api/dps.js:20]: First of all, I'd keep error messages consistent with our other projects. Besides that, it makes sense only if we display them somewhere. For example, by using our notification system.

File: src/assets/icons/LoginIcon.js [hunk 23]
```diff
@@ -0,0 +1,40 @@
+export const LoginIcon = () => (
+  <svg
```
Comment (arturfracala) [inline src/assets/icons/LoginIcon.js:2]: I think that we could optimize this SVG. I'd consider some online tools for this.

**Review by arturfracala — CHANGES_REQUESTED** [review]
We're going in the right direction 👍
